File: backend/java/src/main/java/com/michaelyi/personalwebsite/post/PostService.java
package com.michaelyi.personalwebsite.post;

import com.fasterxml.jackson.core.type.TypeReference;
import com.michaelyi.personalwebsite.cache.CacheService;
import com.michaelyi.personalwebsite.util.StringUtil;
import org.springframework.stereotype.Service;
import org.springframework.web.multipart.MultipartFile;

import java.util.List;
import java.util.NoSuchElementException;

@Service
public class PostService {
        private final PostDao dao;
        private final CacheService cacheService;

        public PostService(PostDao dao, CacheService cacheService) {
                this.dao = dao;
                this.cacheService = cacheService;
        }

        public String createPost(String text, MultipartFile image) {
                Post post;

                try {
                        post = PostUtil.constructPost(text);
                        byte[] imageBytes = PostUtil.getImage(image);
                        post.setImage(imageBytes);
                } catch (IllegalArgumentException e) {
                        throw e;
                }

                Post existingPost = getPost(post.getId());
                if (existingPost != null) {
                        throw new IllegalArgumentException("A post with the same title already exists");
                }

                dao.createPost(post);
                cacheService.delete("getAllPosts");
                return post.getId();
        }

        public Post getPost(String id) {
                if (!StringUtil.isStringValid(id)) {
                        throw new IllegalArgumentException("Id cannot be empty");
                }

                Post post = cacheService.get(String.format("getPost?id=%s", id), Post.class);
                if (post != null) {
                        return post;
                }

                post = dao.getPost(id).orElse(null);
                if (post != null) {
                        String key = String.format("getPost?id=%s", post.getId());
                        cacheService.set(key, post);
                }

                return post;
        }

        public List<Post> getAllPosts() {
                List<Post> posts = cacheService.get("getAllPosts", new TypeReference<List<Post>>() {
                });

                if (posts != null) {
                        return posts;
                }

                posts = dao.getAllPosts();

                if (posts != null) {
                        cacheService.set("getAllPosts", posts);
                }

                return posts;
        }

        public void updatePost(String id, String text, MultipartFile image) {
                if (!StringUtil.isStringValid(id)) {
                        throw new IllegalArgumentException("Id cannot be empty");
                }

                Post post = getPost(id);
                if (post == null) {
                        throw new NoSuchElementException("Post not found");
                }

                Post updatedPost;
                try {
                        updatedPost = PostUtil.constructPost(text);
                } catch (IllegalArgumentException e) {
                        throw e;
                }

                byte[] imageBytes;
                try {
                        imageBytes = PostUtil.getImage(image);
                } catch (IllegalArgumentException e) {
                        imageBytes = null;
                }

                post.setTitle(updatedPost.getTitle());

                if (imageBytes != null) {
                        post.setImage(imageBytes);
                }

                post.setContent(updatedPost.getContent());

                dao.updatePost(post);
                cacheService.delete(String.format("getPost?id=%s", post.getId()));
                cacheService.delete("getAllPosts");
        }

        public void deletePost(String id) {
                if (!StringUtil.isStringValid(id)) {
                        throw new IllegalArgumentException("Id cannot be empty");
                }

                Post post = getPost(id);
                if (post == null) {
                        throw new NoSuchElementException("Post not found");
                }

                dao.deletePost(post.getId());
                String postKey = String.format("getPost?id=%s", post.getId());
                cacheService.delete(postKey);
                cacheService.delete("getAllPosts");
        }
}

File: backend/go/post/post_service.go
package post

import (
	"bytes"
	"encoding/json"
	"fmt"

	"github.com/michaelhyi/personal-website/backend/go/cache"
	"github.com/michaelhyi/personal-website/backend/go/s3"
	"github.com/michaelhyi/personal-website/backend/go/util"
)

type PostService struct {
	Dao          *PostDao
	S3Service    *s3.S3Service
	CacheService *cache.CacheService
}

func (s *PostService) CreatePost(text *string, image []byte) (string, error) {
	post, err := ValidateAndConstructPost(text)

	if err != nil {
		return "", err
	}

	if image == nil || len(image) == 0 {
		return "", fmt.Errorf("Image is required")
	}

	_, err = s.GetPost(&post.ID)

	if err == nil {
		return "", fmt.Errorf("Post already exists")
	}

	s.Dao.CreatePost(&post)

	err = s.S3Service.PutObject(&post.ID, &image)

	if err != nil {
		return "", err
	}

	cacheVal, err := json.Marshal(post)

	if err != nil {
		return "", err
	}

	s.CacheService.Set(fmt.Sprintf("getPost?id=%s", post.ID), string(cacheVal))
	s.CacheService.Delete("getAllPosts")
	return post.ID, nil
}

func (s *PostService) GetPost(id *string) (Post, error) {
	var noop Post

	if util.IsStringInvalid(id) {
		return noop, fmt.Errorf("Invalid id")
	}

	b, err := s.CacheService.Get(fmt.Sprintf("getPost?id=%s", *id))

	if err == nil {
		var post Post
		json.Unmarshal([]byte(b), &post)
	}

	post, err := s.Dao.GetPost(id)

	if err == nil {
		var b []byte
		b, _ = json.Marshal(post)
		s.CacheService.Set(fmt.Sprintf("getPost?id=%s", *id), string(b))
		return post, nil
	}

	return noop, err
}

func (s *PostService) GetPostImage(id *string) ([]byte, error) {
	if util.IsStringInvalid(id) {
		return nil, fmt.Errorf("Invalid id")
	}

	_, err := s.GetPost(id)

	if err != nil {
		return nil, fmt.Errorf("Post not found")
	}

	cachedImg, err := s.CacheService.Get(fmt.Sprintf("getPostImage?id=%s", *id))

	if err == nil {
		return []byte(cachedImg), nil
	}

	image := s.S3Service.GetObject(id)

	if image == nil {
		return nil, fmt.Errorf("Post image not found")
	}

	s.CacheService.Set(fmt.Sprintf("getPostImage?id=%s", *id), string(image))

	return image, nil
}

func (s *PostService) GetAllPosts() ([]Post, error) {
	cachedPosts, err := s.CacheService.Get("getAllPosts")

	var posts []Post
	err = json.Unmarshal([]byte(cachedPosts), &posts)

	if err == nil {
		return posts, nil
	}

	posts, err = s.Dao.GetAllPosts()

	if err != nil {
		return nil, err
	}

	cacheVal, err := json.Marshal(posts)

	if err != nil {
		return nil, err
	}

	s.CacheService.Set("getAllPosts", string(cacheVal))

	return posts, nil
}

func (s *PostService) UpdatePost(id *string, text *string, image []byte) (Post, error) {
	var noop Post
	if util.IsStringInvalid(id) {
		return noop, fmt.Errorf("Invalid id")
	}

	post, err := s.GetPost(id)

	if err != nil {
		return noop, err
	}

	updatedPost, err := ValidateAndConstructPost(text)

	if err != nil {
		return noop, err
	}

	post.Title = updatedPost.Title
	post.Content = updatedPost.Content
	s.Dao.UpdatePost(&post)

	cacheVal, err := json.Marshal(post)

	if err != nil {
		return noop, err
	}

	s.CacheService.Set(fmt.Sprintf("getPost?id=%s", post.ID), string(cacheVal))
	s.CacheService.Delete("getAllPosts")

	if image == nil || len(image) <= 0 {
		return post, nil
	}

	currImage, err := s.GetPostImage(id)

	if err != nil {
		return noop, err
	}

	if image != nil && len(image) > 0 && !bytes.Equal(image, currImage) {
		s.S3Service.DeleteObject(id)
		s.S3Service.PutObject(id, &image)

		s.CacheService.Set(fmt.Sprintf("getPostImage?id=%s", *id), string(image))
	}

	return post, nil
}

func (s *PostService) DeletePost(id *string) error {
	if util.IsStringInvalid(id) {
		return fmt.Errorf("Invalid id")
	}

	_, err := s.GetPost(id)

	if err != nil {
		return err
	}

	s.S3Service.DeleteObject(id)
	s.Dao.DeletePost(id)

	s.CacheService.Delete(fmt.Sprintf("getPost?id=%s", *id))
	s.CacheService.Delete(fmt.Sprintf("getPostImage?id=%s", *id))
	s.CacheService.Delete("getAllPosts")

	return nil
}
